vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+
+/**
+ * @brief User supplied a number that represents the percentage of router 
+ * block swaps relative to the total number of swaps. A random number  
+ * between 0-100 is generated here and compared the user supplied value.
+ * If the random number is less than the user supplied value we indicate
+ * that a router block should be swapped. By doing this we now only swap
+ * router blocks for the percentage of time the user supplied.
+ * 
+ * @param user_supplied_noc_router_swap_percentage The expected number of 
+ * noc router block swaps as a percentage of the total number of swaps
+ * attempted by the placer.
+ * @return true This indicates that a router block should be swapped
+ * @return false This indicates that a router block should not be swapped
+ */
+bool check_for_router_swap(int user_supplied_noc_router_swap_percentage);
+
+/**
+ * @brief Generates a placement move by choosing two rouer cluster blocks to 
+ * swap. First, a random router cluster block is chosen and then another router
+ * cluster block is chosen that cqan be swapped with the initial block such that
+ * the distance travelled by either blocks does not exceed rlim. 
+ * 
+ * @param blocks_affected The two router cluster blocks that are proposed to be
+ * swapped
+ * @param rlim The maximum distance in the x and y direction that a router 
+ * cluster block can travel 
+ * @return e_create_move Result of proposing the move
+ */
+e_create_move propose_router_swap(t_pl_blocks_to_be_moved& blocks_affected, float rlim);
+
+/* Below are functions related to modifying and retreiving the NoC placement stats dastructure*/
+
+/**
+ * @brief Initializes all the stat values to 0 and allocates space for the
+ * number of possible move types offered by the placer. This is needed as
+ * the stats datastructure for the NoC keeps track of which moves caused
+ * a router block to move.
+ * 
+ * @param placer_opts Contains information about all the possible move
+ * types in the placer.
+ */
+void initialize_noc_placement_stats(const t_placer_opts& placer_opts);
+
+/**
+ * @brief Increments a the count of total number of router block moves during
+ * placement. Also increments the move type that caused the router block to
+ * move. This function should be called whenever a noc router block is moved
+ * during placement.
+ * 
+ * @param move_type The type of move in the placer which caused a router block
+ * to move.
+ */
+void update_noc_placement_stats(int move_type);
+
+/**
+ * @brief Displays the NoC placement statistical data.
+ * 
+ */
+void print_noc_placement_stats(void);
+
+/**
+ * @brief Writes out the locations of the router cluster blocks in the
+ * final placement. This information is written out in a format that is
+ * compatible with the RADsim simulator place file. The output of this
+ * function is a .place file.
+ * 
+ * Sample placement file output:
+ *      router_cluster_name layer_number physical_router_id
+ *      ...
```

Comment:
Note from SS to SS: consider if you could figure out a more user friendly name for logical routers on the netlist, that would directly match what Radsim wants.

Reply:
Not sure how to do this right now. Will leave it as future work.